src/algorithms/decentralized.md

```diff
@@ -2,41 +2,40 @@
+## Decentralized SGD
 
-## Decentralized SGD with 
+Now we are going to describe the decentralized SGD algorithm implemented in Bagua. Let's assume the number of workers is $n$, the model parameters on worker $i$ is $\bf x^{(i)}, i \in \{0,...,n-1\}$. Each worker is able to send or receive data directly from any other workers. In each iteration $t$, the algorithm repeats the following steps:
 
-Now we are going to describe the decentralized SGD algorithm implemented in Bagua. Assume the number of workers is $n$, the model parameters on worker $i$ is $\bf x_i$
+1. Each worker $i$ calculate the local gradients of iteration $t$: $\bf g_t^{(i)}$.
+2. Average the local model with its selected peer's model (denote as $x_t^{(j)}$), i.e., $x_{t+\frac{1}{2}}^{(i)} = \frac{x_t^{(i)} + x_t^{(j)}}{2}$
+3. Update the averaged model with the local gradients. $x_{t+1}^{(i)} = x_{t+\frac{1}{2}}^{(i)} - \gamma g_t^{(i)}$.
 
-
-Each worker is able to send or receive data from any other works. The algorithm can be summarized in the following steps:
-
-1. Each worker calculate the local gradients $\bf g_i$
+In step 2, we adopt a strategy to select a peer for each worker in each iteration, such that all workers are properly paired and the data exchanging is efficient in the sense that each worker could exchange data with a different peer between iterations. In short, our strategy evenly split workers into two groups, and dynamically pair workers between two groups, varying from iteration to iteration.
 
 
 ## Communication Overhead
 
-The communication overhead of decentralized SGD is highly related to the degree of network, i.e., the number of connections a worker has to other workers. Different topologies will lead to different degrees of network. For example, in the ring topology, each worker has two connections with its two neighbors; in the random-probing topology, each worker only needs one connection with another random worker. 
-
-Let's assume there are $n$ workers in total. 
-
-
-
-
+The communication overhead of decentralized SGD is highly related to the degree of network, i.e., the number of connections a worker has to other workers. Different topologies or strategies will lead to different degrees of the network. It's obvious that the Decentralized SGD algorithm we described before has a network degree of 1. Therefore, in each iteration, a worker only needs to build one connection with one worker to exchange one time of the model size. We compare the communication complexities of different communication patterns regarding the latency and bandwidth of the busiest node.
 
+| Algorithm     | Latency complexity | Bandwidth complexity  |
+| :-------------: |:-------------:| :-----:|
+| Allreduce (Ring)      | $\mathcal{O}(n)$ | $\mathcal{O}(1)$ |
+| Parameter Server      | $\mathcal{O}(1)$ | $\mathcal{O}(n)$ |
+| Decentralized SGD in Bagua | $\mathcal{O}(1)$ | $\mathcal{O}(1)$ |
 
 
+## Example
 
 
-## Example
+## Benchmark
```

Comment:
which do you think better? leave benchmark in a dedicated page or in each algorithm page?

Reply:
I would prefer a dedicated page for all the benchmarks? Then the tutorial can be focused on the explanation and usage of the algorithm.

Reply:
Then this section contains link to the corresponding section in benchmark page?

Reply:
sure, but I assume we will have one page for all benchmarks, and some benchmarks involved multiple algorithms, then maybe we won't have independent sections for each algorithm. Another option is we just put the benchmark page in a parallel directory as the Algorithm, then it's quite clear to me.